### ISSUE TITLE
ImageViewError with ginga v2.6.1 if the ImageViewCanvas size is not set

### ISSUE THREAD

**Issue by montefra** [ejeschke/ginga#431]
I maintain a PyQt4 tool that uses Ginga to visualize fits files.
With Ginga v2.6.0 the file is correctly visualised while with v2.6.1 I get the following error:

```
Failed to initialize image: new scale would exceed pixel max; scale unchanged
Traceback:
  File "/.../lib/python3.5/site-packages/ginga/ImageView.py", line 783, in _image_set_cb
    self.apply_profile_or_settings(image)
  File "/.../lib/python3.5/site-packages/ginga/ImageView.py", line 840, in apply_profile_or_settings
    self.zoom_fit(no_reset=True)
  File "/.../lib/python3.5/site-packages/ginga/ImageView.py", line 2034, in zoom_fit
    self.center_image(no_reset=no_reset)
  File "/.../lib/python3.5/site-packages/ginga/ImageView.py", line 2319, in center_image
    self.panset_xy(data_x, data_y, no_reset=no_reset)
  File "/.../lib/python3.5/site-packages/ginga/ImageView.py", line 2286, in panset_xy
    self.set_pan(pan_x, pan_y, coord=pan_coord, no_reset=no_reset)
  File "/.../python3.5/site-packages/ginga/ImageView.py", line 2215, in set_pan
    self._reset_bbox()
  File "/.../python3.5/site-packages/ginga/ImageView.py", line 1430, in _reset_bbox
    pan_x, pan_y, win_wd, win_ht)
  File "/.../python3.5/site-packages/ginga/ImageView.py", line 1374, in _calc_bg_dimensions
    raise ImageViewError("new scale would exceed pixel max; scale unchanged")
```

The problem seems to be related to the fact that the latest version cannot correctly get the size of the widget from the image. Manually setting the ``ImageViewCanvas.set_window_size`` to the image size works around the above error.

Here is a minimal example to reproduce the issue. For my test I have used the second fits image in the table in [this website](https://fits.gsfc.nasa.gov/fits_samples.html)

```python
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

import logging
import sys

import ginga.toolkit
ginga.toolkit.use('qt4')
from ginga import AstroImage
from ginga.qtw.ImageViewCanvasQt import ImageViewCanvas
from ginga.misc.Settings import SettingGroup

from PyQt4 import QtGui

app = QtGui.QApplication(sys.argv)

glog = logging.getLogger('ginga')
fn = 'test.fits'
ginga_settings = SettingGroup(logger=glog)

fitsview = ImageViewCanvas(glog, render='widget', settings=ginga_settings)

image = AstroImage.AstroImage(logger=glog)
image.load_file(fn)

# IMPORTANT: uncomment this to get v2.6.1 to work
# fitsview.set_window_size(*image.get_size())

fitsview.set_image(image)
fitsview.image_filename = fn

main_w = QtGui.QMainWindow()
ginga_viewer = fitsview.get_widget()
main_w.setCentralWidget(ginga_viewer)
main_w.show()

sys.exit(app.exec_())
```

**Comment by pllim** [comment]
The image opens fine on reference viewer (i.e., using the `ginga` command to start Ginga). I can reproduce your error on Qt5 using your script (after replacing Qt4 with Qt5 incantations). However, I also notice that in both examples given in `ginga/examples/qt`, there was an explicit call to `fitsview.resize()`, which is missing in your script.

Anyway, just an observation. @ejeschke will know for sure.

**Comment by montefra** [comment]
> The image opens fine on reference viewer (i.e., using the ginga command to start Ginga).
> I can reproduce your error on Qt5 using your script (after replacing Qt4 with Qt5 incantations). 

Thank you for testing this.

> However, I also notice that in both examples given in ginga/examples/qt, there was an explicit call to fitsview.resize(), which is missing in your script.

Yep. To make it work with the v2.6.1 I added:

    fitsview.set_window_size(*image.get_size())

I guess thie is similar to resize().
Actually in the examples (or at least in the first one) there are two resize: one of the ImageViewCanvas and one of the contained widget.

> Anyway, just an observation. @ejeschke will know for sure.

Hope so :smile:

**Comment by ejeschke** [comment]
It seems to be an issue having to do with the default size--I think somewhere along the road to 2.6.1 that changed.  @montefra, if you set the size of your `main_w` do you still get the error?

**Comment by montefra** [comment]
@ejeschke : tomorrow I'll experiment some more and let you know.

**Comment by ejeschke** [comment]
with your test program, I notice that if I simply set 
```python
 fitsview.set_desired_size(1, 1)
```
the widget window is fine (it gets resized to the outer widget).

Now that I think about it a little more, I think the exception is what got added in 2.6.1.  Since the window isn't realized yet at the time of `set_image()` it is trying to calculate the correct size for a 0x0 window.

**Comment by montefra** [comment]
> if you set the size of your main_w do you still get the error?

yes because the canvas knows nothing about the windows it is embedded into it and I add the image before I set the canvas' widget as the CentralWidget

> the widget window is fine (it gets resized to the outer widget).

Same as with me, despite using a different method (I guess that the various ``set_*size`` and ``resize`` are somehow related).

> Now that I think about it a little more, I think the exception is what got added in 2.6.1. Since the window isn't realized yet at the time of set_image() it is trying to calculate the correct size for a 0x0 window.

In version 2.6.0 there is no difference in the output whether I explicitly set the window size or not. So down the line the code behaves (or appears to behave) correctly, as you can see in the figure.
![ginga_comparison](https://cloud.githubusercontent.com/assets/1394820/22587094/74417dfe-ea00-11e6-8aa8-9107e5548eb3.png)

"no_size" means that the size is not explicitly set, "set_size" used the ``set_window_size`` method.
The only side effect that I see in v2.6.0 with no explicit size set is a bunch of log messages:

```
WARNING, scale adjusted downward X (1.0000 -> 0.0000), Y (1.0000 -> 0.0000)                                                                      
ERROR, Scale (0.00) exceeds min scale limit (0.00)
WARNING, scale adjusted downward X (1.0000 -> 0.0000), Y (1.0000 -> 0.0000)
WARNING, scale adjusted downward X (1.0000 -> 0.0000), Y (1.0000 -> 0.0000)
ERROR, Scale (0.00) exceeds min scale limit (0.00)
WARNING, scale adjusted downward X (1.0000 -> 0.0000), Y (1.0000 -> 0.0000)
```

To work around this and avoid getting the log messages, I set now the explicit size.

Would it be preferable to use ``set_window_size`` or ``set_desired_size``?

**Comment by montefra** [comment]
I just made an other small test:

* ``set_window_size``: respect the ``minimumSize`` of the parent window
* ``set_desired_size``: force the underlying widget size to the image size, which might be undesirable (in my case makes the window way too big)

**Comment by ejeschke** [comment]
> set_desired_size: force the underlying widget size to the image size, which might be undesirable (in my case makes the window way too big)

Even in the case where you use (1, 1) for the width and height?

**Comment by montefra** [comment]
If I explicitly set the main window size to (200, 200), this is what happens to the window created by the above script:

 size set in method | ``set_window_size`` | ``set_desired_size``
-|----|-----
image size | respect the main_w size  | window much bigger than (200, 200)
(1, 1) | respect the main_w size with warning | respect the main_w size with warning

Setting (1, 1) respect the main window size, but prints out
```
Error in scaling: resulting scale (0.000941, 0.000941) would result in image size of <1 in width or height
Error in scaling: resulting scale (0.000941, 0.000941) would result in image size of <1 in width or height
```

ps: this is v1.6.1. Do you want a similar test for v1.6.0?

**Comment by ejeschke** [comment]
@montefra, I think you can use whichever method seems best.  I'll take a look at the main issue.

**Comment by montefra** [comment]
@ejeschke : ok, I'll do.

Thank you for Ginga!

**Comment by pllim** [comment]
Is this still an issue for Ginga 2.6.3?

**Comment by montefra** [comment]
I have to test it but I'm far from computers for the next week. I'll let you know asap

**Comment by ejeschke** [comment]
@montefra, I believe this issue is mostly fixed in commit 248353fca586c93cdbe80f3545486ce1167ad98b.  I say "mostly fixed" because you will still get some warnings about "window size undefined" if attempting to set the data prior to the window being realized.  But at least there is no exception thrown that causes the program to abort.

**Comment by ejeschke** [comment]
Closing for now.  We can reopen if the issue reappears.
